Sources/NIOCertificateHotReloading/NIOCertificateHotReloading.swift

```diff
@@ -0,0 +1,279 @@
+        }
+
+        /// This certificate/key can be found at the given filepath.
+        /// - Parameter path: The filepath where the certificate/key can be found.
+        /// - Returns: A `Location`.
+        public static func file(path: String) -> Self { Self(_Backing.file(path: path)) }
+
+        /// This certificate/key is available in memory, and will be provided by the given closure.
+        /// - Parameter provider: A closure providing the bytes for the given certificate or key.
+        /// - Returns: A `Location`.
+        public static func memory(provider: @Sendable @escaping () -> [UInt8]?) -> Self {
+            Self(_Backing.memory(provider: provider))
+        }
+    }
+
+    /// A description of a certificate, in terms of its ``Location`` and ``Encoding``.
+    public struct CertificateDescription: Sendable {
+        public var location: Location
+        public var format: Encoding
+
+        public init(location: Location, format: Encoding) {
+            self.location = location
+            self.format = format
+        }
+    }
+
+    /// A description of a private key, in terms of its ``Location`` and ``Encoding``.
+    public struct PrivateKeyDescription: Sendable {
+        public var location: Location
+        public var format: Encoding
+
+        public init(location: Location, format: Encoding) {
+            self.location = location
+            self.format = format
+        }
+    }
+
+    private struct CertificateKeyPair {
+        var certificate: NIOSSLCertificateSource
+        var privateKey: NIOSSLPrivateKeySource
+    }
+
+    private let refreshInterval: TimeAmount
+    private let certificateDescription: CertificateDescription
+    private let privateKeyDescription: PrivateKeyDescription
+    private let state: NIOLockedValueBox<CertificateKeyPair?>
+
+    /// A `NIOSSLContextConfigurationOverride` that will be used as part of the NIO application's TLS configuration.
+    /// Its certificate and private key will be kept up-to-date via the reload mechanism the ``TimedCertificateReloader``
+    /// implementation provides.
+    public var sslContextConfigurationOverride: NIOSSLContextConfigurationOverride {
+        get {
+            var override = NIOSSLContextConfigurationOverride()
+            guard let certificateKeyPair = self.state.withLockedValue({ $0 }) else {
+                return override
+            }
+            override.certificateChain = [certificateKeyPair.certificate]
+            override.privateKey = certificateKeyPair.privateKey
+            return override
+        }
+    }
+
+    /// Initialize a new ``TimedCertificateReloader``.
+    /// - Parameters:
+    ///   - refreshInterval: The interval at which attempts to update the certificate and private key should be made.
+    ///   - certificateDescription: A ``CertificateDescription``.
+    ///   - privateKeyDescription: A ``PrivateKeyDescription``.
+    public init(
+        refreshingEvery refreshInterval: TimeAmount,
```

Comment:
Why are we not taking a `Duration` here?

Reply:
Because it's only available on macOS 13+. Using `TimeAmount` we can support 11+. I was on the fence about this.

Reply:
I see. Can we add a second init with duration then that has an availability annotation?

Reply:
Yeah good callout.